torch/csrc/utils/auto_gpu.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+// RAII structs to set CUDA device
+
+#include <string>
+#include <stdexcept>
+
+#ifdef WITH_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#endif
+
+struct AutoGPU {
+  explicit AutoGPU(int device=-1) {
+    setDevice(device);
+  }
+
+  ~AutoGPU() {
+#ifdef WITH_CUDA
+    if (original_device != -1) {
+      cudaSetDevice(original_device);
+    }
+#endif
+  }
+
+  inline bool setDevice(int device) {
+#ifdef WITH_CUDA
+    if (device == -1) {
+      return false;
+    }
+    if (original_device == -1) {
+      cudaCheck(cudaGetDevice(&original_device));
+      if (device != original_device) {
+        cudaCheck(cudaSetDevice(device));
+      }
+    } else {
+      cudaCheck(cudaSetDevice(device));
```

Comment:
`if device != original_device`? We're doing **a lot** of these calls, so I think we should try to save them.